### PR TITLE
[#17727] Implement Dropdown input component

### LINKED ISSUE #17727: Quo2 - Implement Dropdown input component
designs:
https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=26956-36369&mode=design&t=SkhycMIt3idQHKAr-4

Acceptance criteria -
-align with figma props and best practices
- component spec test
- all themes and variants

### PR DESCRIPTION
**Pull request by tumanov-alex**
fixes #17727 

### Summary

# Implement dropdown-input component based on the original dropdown one

[Figma](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?node-id=26956%3A25449&mode=dev)
| --- | --- | --- |

### CODE REVIEW

**Comment by flexsurfer** [comment]
just wondering how's it different from `quo.components.dropdowns.dropdown.view`

File: src/quo/components/dropdowns/dropdown_input/style.cljs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  [theme blur?]
+  {:color         (if blur?
+                    (colors/theme-colors colors/neutral-50 colors/neutral-40
+                                         theme)
+                    (colors/theme-colors colors/neutral-80-opa-40
+                                         colors/white-opa-70 theme))
+   :margin-bottom gap})
+
+(def root-container
+  {:width dropdown-width})
```
Comment (flexsurfer) [inline src/quo/components/dropdowns/dropdown_input/style.cljs]: empty line

**Comment by tumanov-alex** [comment]
@flexsurfer i know what you mean — the diff between this component and the original dropdown is thin, it mostly about different use cases. Nevertheless team decided that we should have a separate dropdown-input component for this particular case and that it would be better not to re-use the original dropdown.

is that correct? @mariocnovoa 
cc @J-Son89

**Comment by J-Son89** [comment]
Yep, we follow the hierarchy that is set in figma. Otherwise we try to get changes shoehorned into components and especially if the designs change. 
This happened previously with the button and drop-down. The button component ended up with a bunch of hacks to make the drop-down ui work with it and the button was difficult to maintain/update/use.

If there is some issues with this approach it's best to discuss with designers to update their usage of the components. The less components we have to support the better 👌

**Comment by tumanov-alex** [comment]
@J-Son89 initially i've had concerns about copy-pasting dropdown code (its not dry i thought) but when you consider how difficult it would be to support a component that's basically 1.5 components, you realize it's better to have two simple modules that one, but complex and bug prone.

**Review by ilmotta — APPROVED** [review]
@tumanov-alex, @J-Son89, I'm not too concerned about the component duplication because the root of the problem is in the design system (on the Figma layer).

Our design system is full of "almost identical" things, full of one usage components. The dropdown is not the only one (e.g. selectors and tags components). It would be way better for maintainability, especially at MVP stage if designers (and devs in turn) judiciously applied something like the atomic design methodology.

I'm not saying we can't do anything to keep things DRY, of course we can, but I think in the end it's an uphill battle. And changing that landscape would require a complete design shift, both in mindset and in the actual components that would be built as legos.

But having said that, it's unclear to me if we shouldn't at least reuse some parts between both components. I'm approving anyway because I don't think it's worth refactoring to be more DRY if you already have something working.

**Comment by J-Son89** [comment]
> @tumanov-alex, @J-Son89, I'm not too concerned about the component duplication because the root of the problem is in the design system (on the Figma layer).
> 
> 
> 
> Our design system is full of "almost identical" things, full of one usage components. The dropdown is not the only one (e.g. selectors and tags components). It would be way better for maintainability, especially at MVP stage if designers (and devs in turn) judiciously applied something like the atomic design methodology.
> 
> 
> 
> I'm not saying we can't do anything to keep things DRY, of course we can, but I think in the end it's an uphill battle. And changing that landscape would require a complete design shift, both in mindset and in the actual components that would be built as legos.
> 
> 
> 
> But having said that, it's unclear to me if we shouldn't at least reuse some parts between both components. I'm approving anyway because I don't think it's worth refactoring to be more DRY if you already have something working.
> 
> 

@ilmotta, I agree. The issue is at the source (figma). Indeed it is too late to change what is already in place. We should however bring this up with designers as they go to add new components/designs. Let's raise this as a point of discussion with them as I believe browser is currently being designed.

File: src/quo/components/dropdowns/dropdown_input/style.cljs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+(ns quo.components.dropdowns.dropdown-input.style
+  (:require
+    [quo.foundations.colors :as colors]))
+
+(def gap 8)
+(def dropdown-width 288)
```
Comment (J-Son89) [inline src/quo/components/dropdowns/dropdown_input/style.cljs]: I think we don't want to set width. It's better to use flex and then we set margin on the screen where the component is being used.
Usually via a 'container-style' prop which is passed to the outer container

File: src/quo/components/dropdowns/dropdown_input/view.cljs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+(defn- view-internal
+  [{:keys [state theme on-press icon?
+           label? blur?
+           accessibility-label]
+    :or   {state  :default
+           icon?  true
+           label? true
+           blur?  false}
+    :as   props}
+   label header-label]
```
Comment (J-Son89) [inline src/quo/components/dropdowns/dropdown_input/view.cljs]: Any particular reason label and header label aren't passed in the map?

File: src/quo/core.cljs [hunk 4]
```diff
@@ -232,6 +233,7 @@
 ;;;; Dropdowns
 (def dropdown quo.components.dropdowns.dropdown.view/view)
 (def network-dropdown quo.components.dropdowns.network-dropdown.view/view)
+(def dropdown-input quo.components.dropdowns.dropdown-input.view/view)
```
Comment (J-Son89) [inline src/quo/core.cljs:236]: Can we keep this alpha sorted, makes this file easier to keep organised/ check what's there 👍

**Review by J-Son89 — APPROVED** [review]
Nice one @tumanov-alex! 🚀

**Comment by ilmotta** [comment]
> > @tumanov-alex, @J-Son89, I'm not too concerned about the component duplication because the root of the problem is in the design system (on the Figma layer).
> > Our design system is full of "almost identical" things, full of one usage components. The dropdown is not the only one (e.g. selectors and tags components). It would be way better for maintainability, especially at MVP stage if designers (and devs in turn) judiciously applied something like the atomic design methodology.
> > I'm not saying we can't do anything to keep things DRY, of course we can, but I think in the end it's an uphill battle. And changing that landscape would require a complete design shift, both in mindset and in the actual components that would be built as legos.
> > But having said that, it's unclear to me if we shouldn't at least reuse some parts between both components. I'm approving anyway because I don't think it's worth refactoring to be more DRY if you already have something working.
> 
> @ilmotta, I agree. The issue is at the source (figma). Indeed it is too late to change what is already in place. We should however bring this up with designers as they go to add new components/designs. Let's raise this as a point of discussion with them as I believe browser is currently being designed.

Let's do it 👍🏼 The important part for me is that we (but especially designers) understand the first principles behind this problem and how they propagate to code as inefficiencies and bugs. We just need to make sure the discussion doesn't focus on only one or two particular examples, which is what usually happens.

**Comment by J-Son89** [comment]
Probably it would help if some developers could review the components before they're "finalised" too

File: src/quo/components/dropdowns/dropdown_input/view.cljs (hunk 3 above)
Comment (tumanov-alex) [reply src/quo/components/dropdowns/dropdown_input/view.cljs]: to be honest i thought we have a convention to pass such info as labels outside of maps to imitate api of core components like rn/text, where you also pass label outside of map

Comment (J-Son89) [reply src/quo/components/dropdowns/dropdown_input/view.cljs]: yeah that's true, I think I had not seen any component with 2 args like this, i.e `[map arg arg]` but most components like button, dropdown etc take the form of `[map arg]`. 
I don't. think it matters so much, I wonder for consistency is it better to put header-label in the map instead? how does the other dropdown handle it, let's match that if it has a similar header-label?

Comment (tumanov-alex) [reply src/quo/components/dropdowns/dropdown_input/view.cljs]: done 👍

File: src/quo/core.cljs (hunk 4 above)
Comment (tumanov-alex) [reply src/quo/core.cljs:236]: for some reason i get no error from lint about this 🤔

File: src/quo/components/dropdowns/dropdown_input/style.cljs (hunk 2 above)
Comment (tumanov-alex) [reply src/quo/components/dropdowns/dropdown_input/style.cljs]: if you mean setting flex 1 to `root-container` & `container`, then we'll have an issue with height also being increased as component will grow. I've used width 100% combined with height 40px, if that's ok

File: src/quo/components/dropdowns/dropdown_input/view.cljs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+           label? blur?
+           accessibility-label header-label]
+    :or   {state        :default
+           icon?        true
+           label?       true
+           blur?        false
+           header-label "Label"}
+    :as   props}
+   label]
+  (let [icon-size      20
```
Comment (smohamedjavid) [inline src/quo/components/dropdowns/dropdown_input/view.cljs]: We can move this value to the properties ns/file and use it from there.

**Review by smohamedjavid — APPROVED** [review]
Nice work! @tumanov-alex 🚀

**Review by Francesca-G — APPROVED** [review]
Everything looks good besides the label color, please check it 🙏
here's the [review](https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS/Feedback-for-Mobile?type=design&node-id=3374%3A311523&mode=design&t=IGDHIH3hD8zPKCth-1)

**Pull request merged**